Pgadmin update 6.8, fix build breakage on master

gador opened the pull request:
###### Description of changes

Update to the [new release](https://www.pgadmin.org/docs/pgadmin4/latest/release_notes_6_8.html).
This fixes a regression tracked in #169322.

Unfortunately, the recent update of `flask` and `werkzeug` cause a build failure.

This makes the override necessary. Since pgadmin4 isn't imported anywhere else (a leaflet), it shouldn't cause inconsistent python environments.

I haven't figured out a way to nicely import the different pythonScope in the test suite for the regression testing and had to do with `overlays`.

Thanks @FRidh for the pointer to `overrideScope` ! (Maybe it is worthwhile adding that to the manual..) and to @teto for the initial bug report.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

gador committed:
pgadmin4: 6.7 -> 6.8

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

FRidh committed:
pgadmin: fix pgadmin4 command

gador committed:
pgadmin: add werkzeug override

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
pgadmin4: make regression test use the same packages

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

mweinelt commented:
Apparently pgadmin4 6.7 built without these overrides over in #169434.

gador commented:
Yes. It builds. But it doesn't start and the tests in nixos fail.

Due to the nature of the tests, they cannot be included as tests in the main package..

FRidh commented:
> Thanks @FRidh for the pointer to overrideScope ! (Maybe it is worthwhile adding that to the manual..) and to @teto for the initial bug report.

Actually, it's not there because it doesn't work with `buildEnv/withPackages`. I see you worked around that now.

FRidh commented on nixos/tests/pgadmin4.nix:
```diff
@@ -5,62 +5,93 @@ import ./make-test-python.nix ({ pkgs, lib, ... }:
     nodes.machine = { pkgs, ... }: {
       imports = [ ./common/x11.nix ];
+      # needed because pgadmin 6.8 will fail, if those dependencies get updated
+      nixpkgs.overlays = [
```
overriding is already done in the derivation, so why do it here as well

gador replied:
See the comment below.

1) Pgadmin is not wrapped with test only python packages

2) Pgadmin (the app) is wrapped, but not the tests. To run the python tests, a python environment is needed.

SuperSandro2000 replied:
Please move this into the package possible behind a flag which builds it with test dependencies.

gador replied:
This is already in the package in L74ff (`pgadmin4`). The override in the python package set need to be in the package, as well as here in the test. (Or do you know any way to re-use the `overrideScope` between derivations?)
 
Adding a test-flag to the pgadmin4 derivation is more work and more confusing than just adding the two test packages here in L71-L72 (`nixos/tests/pgadmin4`).

SuperSandro2000 replied:
If the overrideScope is required pgadmin should be not in pythonPackages and do them there. This will definitely bit rot over time and be forgotten when updating packages.

gador replied:
What else would you suggest?
pgadmin will not start with the newer packaged versions and needs the older versions. Even if the overrides bit-rots and I forget about the overrides here: pgadmin is a leaflet packag and the tests will catch any build- and runtime errors due to the then-stale overrides.

SuperSandro2000 replied:
They need to be moved to ``pkgs/tools/admin/pgadmin/default.nix`` and only be in that file.

gador replied:
So, if I understand you correctly, the `nixpkgs.overlays` must be removed here from the test and the `overrideScope` from the pgadmin.nix derivation must be imported here?

I really wrestled a long time with this (see https://github.com/NixOS/nixpkgs/pull/169370#discussion_r855898902) and this is the only way I got it to work. Any pointers of including the `overrideScope` from the pgadmin.nix into this test environment would be greatly appreciated.

gador replied:
Ok, apparently I work better when its not late at night. I gave it a try to pass the python environment to the test from the derivation. Id appreciate a second (or third, forth...) look ;-)

SuperSandro2000 replied:
> So, if I understand you correctly, the `nixpkgs.overlays` must be removed here from the test and the `overrideScope` from the pgadmin.nix derivation must be imported here?

I took a second, deeper look and how it is right now not necessarily but if you would run the test standalone it wouldn't use the python env from pgadmin, so we still need to change it.

I think the easiest way would be to passthru the right things from pgadmin and use them in the test.

gador replied:
Ok, I think I did that. 

The testcall to standalone is left as it is. The imported pgadmin package there is the one with the built-in python overrides.
The testcall to package is changed to pass the python environment. 

This way we only have to change it in one place.

gador replied:
@SuperSandro2000
Do you have time for (yet) another look?

FRidh commented on nixos/tests/pgadmin4.nix:
```diff
@@ -5,62 +5,93 @@ import ./make-test-python.nix ({ pkgs, lib, ... }:
-        python-with-needed-packages
         chromedriver
         chromium
+        (python3.withPackages
```
why is this environment needed? pgadmin should be wrapped with it

gador replied:
Yes, Pgadmin is wrapped.
But the tests (which are executed here, because they need a postgres server running) are not.
Also, Pgadmin is not wrapped with test python packages (such as testscenarios). 

I tried all kinds of things, but this is the only working solution I have found, which overrides the packages.
If you have any suggestions of how else to implement the regression tests, I'm open to them 🙂

FRidh replied:
It's probably cleaner to put take the various *inputs then from `pgadmin4` and build a environment using those. No way to repeat packages that way.

gador replied:
Yes, i totally agree. Unfortunately, I can't seem to get it to work.

What I tried:
- adding `pkgs.pgadmin4.propagatedBuildInputs` to  `environment.systemPackages = with pkgs; [ ]`
- adding `pkgs.pgadmin4.propagatedBuildInputs` to `(python3.withPackages (ps: with pythonPackages; [ ] `
- trying to instigate a `nix-shell  -p pgadmin` within the VM (which would have all those dependencies linked)

At this point I am at a loss. I really don't like repeating the packages from the `pgadmin4` derivation, on the other hand 
this is the only way I got it to work.
I have looked at other `nixos/tests` examples but didn't find anything related.

Do you have any suggestions?

gador replied:
I found a way.

gador committed:
pgadmin4: fix tests

this commit passes the build dependencies to the
pgadmin nixos test for package and regression testing.

Also added changelog and some clarifying comments.

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador commented:
@FRidh could you take another look? Thanks :smile:

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/833

SuperSandro2000 commented on pkgs/tools/admin/pgadmin/default.nix:
```diff
@@ -1,21 +1,23 @@
 
   pname = "pgadmin";
-  version = "6.7";
+  majorVersion = "6";
+  minorVersion = "8";
+  version = "${majorVersion}.${minorVersion}";
```
Please revert that. We have ``lib.versions.majorMinor``, ``lib.versions.major`` and ``lib.versions.minor``.

SuperSandro2000 commented on pkgs/tools/admin/pgadmin/default.nix:
```diff
@@ -25,9 +27,78 @@ let
+        "test_aborting"
+      ];
+    });
+    flask-paranoid = prev.flask-paranoid.overridePythonAttrs (oldAttrs: rec {
+      # Nothing of interest changed from 0.2 to 0.3
+      doCheck = false;
+    });
```
```suggestion
```
If the tests are failing we should fix it in the package itself.

gador replied:
The package tests work fine on their own. Just in combination with the `flask` and `werkzeug` downgrade they fail here.
All 6 tests fail, so we just disable them here.

gador committed:
pgadmin: revert version string

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
pgadmin4: pass pythonEnv as variable

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

gador committed:
pgadmin: minor bugfix in version string

Signed-off-by: Florian Brandes <florian.brandes@posteo.de>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/pgadmin4-fails-to-start/19004/4
